lib/teracy-dev/location/git_synch.rb

```diff
@@ -213,21 +235,190 @@ def check_branch(current_ref, desired_branch)
+              repo_username_key, repo_password_key
+          else
+            # if credentials are not exists then inform user about it
+
+            # we has internet but unable to connect
+            if error_msg.to_s.match('fatal: unable to access')
+              fail_to_pull remote_url
+            # we has no internet or has no credentials
+            else
+              internet_down_or_credential_not_found_to_pull remote_url,
+                repo_username_key, repo_password_key
+            end
+          end
+        end
+
+        return false
+      end
+
+      # return true if clone success, false otherwise
+      def attempt_to_clone_using_http_auth? remote_url, dir
+        processed_remote_url, credential_exists,
+        repo_username_key, repo_password_key = get_remote_credentials remote_url
+
+        # if it is not https format
+        # we will assume it is ssh format and use normal clone
+        if processed_remote_url.nil?
+          clone_success = git_clone remote_url, dir
+
+          return true if clone_success
+
+          fail_to_access remote_url
+        else
+          # otherwise
+          # it is https url
+          # and wheter it has credentials or not, still clone it
+
+          @logger.info("Attempting to using your configurated credentials for #{remote_url}")
+
+          clone_success, error_msg = git_clone processed_remote_url, dir
+
+          return true if clone_success
+
+          # we fail, so raise corresponding messages
+
+          # we has credentials
+          if credential_exists
+            credentials_are_present_but_fail_to_access remote_url,
+                repo_username_key, repo_password_key
+          else
+            # we has internet but unable to connect
+            if error_msg.match('fatal: unable to access')
+              fail_to_access remote_url
+            # we has no internet or has no credentials
+            else
+              internet_down_or_credential_not_found_to_access remote_url,
+                repo_username_key, repo_password_key
+            end
+          end
+        end
+
+        return false
+      end
+
+      def get_remote_credentials remote_url
+        matches = remote_url.match(/(https?):\/\/(.*)/)
+
+        return nil if matches.nil?
+
+        repo_host = Util.get_hostname matches[2]
+
+        # find from ENV if there are any of these credentials
+        # example: GITHUB_USERNAME, GITHUB_PASSWORD
+        repo_username_key = "#{repo_host}_username".upcase
+
+        repo_password_key = "#{repo_host}_password".upcase
+
+        credential_exists = !ENV[repo_username_key].nil? and !ENV[repo_password_key].nil?
+
+        @logger.debug("repo_username: #{repo_username_key}=#{ENV[repo_username_key]}, repo_password_key: #{repo_password_key}=#{ENV[repo_password_key]}")
+
+        if credential_exists
+          remote_url = "#{matches[1]}://#{ENV[repo_username_key]}:'#{ENV[repo_password_key]}'@#{matches[2]}"
```

Comment:
@phuonglm special password will me quoted here

Reply:
```
fail to clone:
Cloning into 'teracy-dev-iorad-common'...
fatal: unable to access 'https://phuonglm:031@4107@github.com/iorad/teracy-dev-iorad-common.git/': Could not resolve host: 4107@github.com
```
quote is not enough